src/core/mainloop/mainloop.c

```diff
@@ -2003,6 +2005,55 @@ add_entropy_callback(time_t now, const or_options_t *options)
   return ENTROPY_INTERVAL;
 }
 
+/** Periodic callback: if there has been no network usage in a while,
+ * enter a dormant state. */
+static int
+check_network_participation_callback(time_t now, const or_options_t *options)
+{
+  /* If we're a server, we can't become dormant. */
+  if (server_mode(options)) {
+    note_user_activity(now);
+    goto end;
+  }
+
+  /* If we're running an onion service, we can't become dormant. */
+  /* XXXX this would be nice to change, so that we can be dormant with a
+   * service. */
+  if (hs_service_get_num_services() || rend_num_services()) {
+    note_user_activity(now);
+    goto end;
+  }
+
+  /* XXXX Add an option to never become dormant. */
+
+  /* If we have any currently open entry streams other than "linked"
+   * connections used for directory requests, those count as user activity.
+   */
+  /* XXXX make this configurable? */
+  if (connection_get_by_type_nonlinked(CONN_TYPE_AP) != NULL) {
+    note_user_activity(now);
+    goto end;
```

Comment:
Personal preferences so do it has you please but I would probably make this function a bit different in terms of flow.

I would use a label that says "not_dormant" which would note the user activity and finish the function.

The point of this would be to not copy the same code everywhere when we are not dormant, have every condition fall to the `not_dormant` state and if you reach the end of the function (pass through all conditions) you are dormant.

It would, imo, also helps understand the function logic and sequential flow instead of having different condition possibly doing different things which is error-prone imo when adding more stuff later.

Reply:
8e90d1d9794754ee323dc5fa38837b0f41ef69e0 tried to do it this way -- better?